src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs

```diff
@@ -0,0 +1,377 @@
+
+        /// <summary>
+        /// The type of filesystem object represented by this entry.
+        /// </summary>
+        public TarEntryType EntryType => _header._typeFlag;
+
+        /// <summary>
+        /// The ID of the group that owns the file represented by this entry.
+        /// </summary>
+        /// <remarks>This field is only supported in Unix platforms.</remarks>
+        public int Gid
+        {
+            get => _header._gid;
+            set => _header._gid = value;
+        }
+
+        /// <summary>
+        /// A timestamps that represents the last time the contents of the file represented by this entry were modified.
+        /// </summary>
+        /// <remarks>In Unix platforms, this timestamp is commonly known as <c>mtime</c>.</remarks>
+        public DateTimeOffset ModificationTime
+        {
+            get => _header._mTime;
+            set
+            {
+                if (value < DateTimeOffset.UnixEpoch)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value));
+                }
+                _header._mTime = value;
+            }
+        }
+
+        /// <summary>
+        /// When the <see cref="EntryType"/> indicates an entry that can contain data, this property returns the length in bytes of such data.
+        /// </summary>
+        /// <remarks>The entry type that commonly contains data is <see cref="TarEntryType.RegularFile"/> (or <see cref="TarEntryType.V7RegularFile"/> in the <see cref="TarFormat.V7"/> format). Other uncommon entry types that can also contain data are: <see cref="TarEntryType.ContiguousFile"/>, <see cref="TarEntryType.DirectoryList"/>, <see cref="TarEntryType.MultiVolume"/> and <see cref="TarEntryType.SparseFile"/>.</remarks>
+        public long Length => _header._dataStream != null ? _header._dataStream.Length : _header._size;
+
+        /// <summary>
+        /// When the <see cref="EntryType"/> indicates a <see cref="TarEntryType.SymbolicLink"/> or a <see cref="TarEntryType.HardLink"/>, this property returns the link target path of such link.
+        /// </summary>
+        public string LinkName
+        {
+            get => _header._linkName;
+            set
+            {
+                if (_header._typeFlag is not TarEntryType.HardLink and not TarEntryType.SymbolicLink)
+                {
+                    throw new NotSupportedException(SR.TarEntryHardLinkOrSymLinkExpected);
+                }
+                _header._linkName = value;
+            }
+        }
+
+        /// <summary>
+        /// Represents the Unix file permissions of the file represented by this entry.
+        /// </summary>
+        /// <remarks>The value in this field has no effect on Windows platforms.</remarks>
+        public TarFileMode Mode
+        {
+            get => (TarFileMode)_header._mode;
+            set
+            {
+                if ((int)value is < 0 or > 4095) // 4095 in decimal is 7777 in octal
```

Comment:
Interesting. I wonder what makes it more efficient to cast rather than check the value. I'll make the change.

Reply:
@gfoidl would https://github.com/dotnet/runtime/pull/67448 eliminate this one too?

Reply:
No, the cast is required here, as it actually changes semantics of the comparison.

Reply:
>  I wonder what makes it more efficient to cast rather than check the value. I'll make the change.

It's removing a branch, doing one comparison instead of a comparison, a branch, and a second comparison. Unfortunately today neither Roslyn nor the JIT optimize such a range check in this manner.

Reply:
https://github.com/dotnet/runtime/pull/67448 is unrelated, as here are no bound-checks involved.

> Unfortunately today neither Roslyn nor the JIT optimize such a range check in this manner.

https://github.com/dotnet/roslyn/issues/60534 tracks this.

Reply:
Thanks for the explanations. No change made.